libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,90 @@
+#include "adc.h"
+#include "stm32f0xx.h"  
+
+#include <stdio.h>
+
+uint32_t prv_get_channel(GPIOAddress* address) {
+  volatile uint8_t adc_channel = address->pin;
+
+  switch (address->port) {
+     case GPIO_PORT_A:
+       if (address->pin > 7) { return -1; }
+       break;
+
+     case GPIO_PORT_B:
+       if (address->pin > 1) { return -1; }
+       adc_channel += 8;
+       break;
+
+     case GPIO_PORT_C:
+       if (address->pin > 5) { return -1; }
+       adc_channel += 10;
+       break;
+
+     default:
+       return -1;
+  }
+
+  return adc_channel;
+}
+
+static uint16_t prv_adc_read_single(GPIOAddress* address, uint8_t adc_channel) {
+  // In single mode, the ADSTART bit must be explicitly set
+  ADC1->CR ^= ADC_CR_ADSTART;
+  ADC1->CHSELR = 1 << adc_channel;
```

Comment:
By setting CHSELR, you only allow a single channel to be converted. Also, you aren't waiting for the conversion to complete so you might be getting old data.